### ISSUE TITLE
Tpetra::MultiVector::Dot() slower than Epetra for short vectors, faster for long vectors

### ISSUE THREAD

**Issue by CamelliaDPG** [trilinos/Trilinos#1455]
The driver pasted below allows CLI arguments to control the length of two vectors that are dotted together, as well as the number of such dot products that should be done.  It does this with both `Tpetra::MultiVector` and `Epetra_Vector` objects, with a serial distribution, and reports the cumulative time for each of Epetra and Tpetra.  Here are the timing results from a run on a shared Intel Xeon E7-4880 machine (times in seconds):

numDots | globalRowCount | Epetra | Tpetra | Slowdown
--------- | -----------------| -------| --------| --------
1000000 | 50 | 0.43 | 8.85 | 21x
100000 | 500 | 0.11 | 2.86 | 26x
10000 | 5000 | 0.07 | 2.25 | 32x
1000 | 50000 | 0.08 | 2.20 | 28x
100 | 500000 | 0.07 | 2.17 | 31x
10 | 5000000 | 0.15 | 2.29 | 15x
1 | 50000000 | 0.15 | 2.29 | 15x

Can this be addressed?  From some profiling, it does appear that part of the story involves calls to `getVector()`, which result in new Tpetra Vector objects being created in each dot product (a shallow copy).  That is probably why the slowdown is 15x as the vectors get larger, but as high as 32x for smaller vectors: with the smaller vectors you see the cost of the object creation and destruction.  It appears, though, that the calls to `getVector()` are only about half the story: there is still a 15x slowdown to account for.

```C++
// Epetra
#include <Epetra_Map.h>
#include <Epetra_SerialComm.h>
#include <Epetra_Vector.h>

// Tpetra
#include <Tpetra_MultiVector.hpp>

// Teuchos
#include <Teuchos_Comm.hpp>
#include <Teuchos_RCP.hpp>
#include <Teuchos_TimeMonitor.hpp>

#include <iostream>
#include <fstream>
#include <string>

using namespace std;
using namespace Teuchos;

typedef long long    GlobalOrdinal; // signed types preferred
typedef int          LocalOrdinal;  // signed types preferred
typedef double Scalar;
typedef Tpetra::Map<LocalOrdinal, GlobalOrdinal>::node_type                    NodeType;
typedef Tpetra::Map<LocalOrdinal,GlobalOrdinal,NodeType>                       TpetraMap;
typedef Tpetra::MultiVector<Scalar,LocalOrdinal,GlobalOrdinal, NodeType>       TpetraMultiVector;

int main(int argc, char *argv[])
{
  int numDots = 10000;
  
  int globalRowCount = 1500;
  
  if (argc == 2)
  {
    numDots = atoi(argv[1]);
    cout << "setting numDots to " << numDots << endl;
  }
  
  if (argc == 3)
  {
    numDots = atoi(argv[1]);
    globalRowCount = atoi(argv[2]);
  }
  
  cout << "numDots is " << numDots << endl;
  cout << "globalRowCount is " << globalRowCount << endl;
  
  RCP<Time> epetraTimer = TimeMonitor::getNewCounter("Epetra dot");
  RCP<Time> tpetraTimer = TimeMonitor::getNewCounter("Tpetra dot");
  
  /*** Epetra setup ***/
  Epetra_SerialComm serialComm;
  Epetra_Map serialMap(globalRowCount, 0, serialComm);
  
  RCP<Epetra_Vector> a_epetra = rcp( new Epetra_Vector(serialMap) );
  RCP<Epetra_Vector> b_epetra = rcp( new Epetra_Vector(serialMap) );
  
  for (int localOrdinal=0; localOrdinal<serialMap.NumMyElements(); localOrdinal++)
  {
    (*a_epetra)[localOrdinal] = 2.0;
    (*b_epetra)[localOrdinal] = 0.5;
  }
  
  /*** Tpetra setup ***/
  RCP<Teuchos::Comm<int>> serialCommTeuchos = Teuchos::rcp( new Teuchos::SerialComm<int>() );
  RCP<const TpetraMap> serialMapTpetra = rcp( new TpetraMap(globalRowCount, 0, serialCommTeuchos));
  RCP<TpetraMultiVector> a_tpetra = rcp( new TpetraMultiVector(serialMapTpetra, 1)); // numVectors = 1
  RCP<TpetraMultiVector> b_tpetra = rcp( new TpetraMultiVector(serialMapTpetra, 1)); // numVectors = 1
  
  auto a_localView = a_tpetra->getLocalView<Kokkos::HostSpace>();
  auto b_localView = b_tpetra->getLocalView<Kokkos::HostSpace>();
  
  for (int localOrdinal=serialMapTpetra->getMinLocalIndex(); localOrdinal<=serialMapTpetra->getMaxLocalIndex(); localOrdinal++)
  {
    a_localView(localOrdinal,0) = 2.0;
    b_localView(localOrdinal,0) = 0.5;
  }
  
  for (int i=0; i<numDots; i++)
  {
    double epetraResult;
    int err;
    epetraTimer->start();
    epetraTimer->incrementNumCalls();
    err = a_epetra->Dot(*b_epetra, &epetraResult);
    epetraTimer->stop();

    if (err != 0)
    {
      cout << "Epetra dot returned err code " << err << ".  Halting execution.\n";
      exit(err);
    }
    
    std::vector<Scalar> tpetraResult( 1 );
    tpetraTimer->start();
    tpetraTimer->incrementNumCalls();
    a_tpetra->dot(*b_tpetra, tpetraResult);
    tpetraTimer->stop();
  }
  
  TimeMonitor::summarize(serialCommTeuchos.ptr());
  
  return 0;
}
```

**Comment by CamelliaDPG** [comment]
@trilinos/tpetra

**Comment by mhoemmen** [comment]
Could you try the Tpetra::Vector::dot function (that returns the value) and see if you get the same performance results? thanks!

**Comment by CamelliaDPG** [comment]
Sure!  Here are the results with `Tpetra::Vector::dot`.

numDots | globalRowCount | Epetra | Tpetra | Slowdown
--------- | -----------------| -------| --------| --------
1000000 | 50 | 0.35 | 9.13 | 26x
100000 | 500 | 0.14 | 2.81 | 20x
10000 | 5000 | 0.05 | 2.25 | 45x
1000 | 50000 | 0.06 | 2.20 | 37x
100 | 500000 | 0.07 | 2.14 | 31x
10 | 5000000 | 0.14 | 2.22 | 16x
1 | 50000000 | 0.14 | 2.25 | 16x

**Comment by mhoemmen** [comment]
@CamelliaDPG Could you post compiler and build options?

**Comment by CamelliaDPG** [comment]
@mhoemmen Here's my do-configure script.  It looks like the particular build I was using for the timings was from 5/16, probably the develop branch.

```
EXTRA_ARGS=$@

rm -f CMakeCache.txt

module load sems-env
module load sems-gcc/6.1.0
module load sems-openmpi/1.8.7
module load sems-parmetis/4.0.3/parallel
module load sems-hdf5/1.8.12/parallel
module load sems-netcdf/4.3.2/parallel
module load sems-superlu/4.3/base

export NET_CDF_DIR=/projects/sems/install/rhel6-x86_64/sems/tpl/netcdf/4.3.2/gcc/6.1.0/openmpi/1.8.7/parallel 
export PARMETIS_DIR=/projects/sems/install/rhel6-x86_64/sems/tpl/parmetis/4.0.3/gcc/6.1.0/openmpi/1.8.7/parallel
export SUPERLU_DIR=/projects/sems/install/rhel6-x86_64/sems/tpl/superlu/4.3/gcc/6.1.0/base

cmake \
-DCTEST_USE_LAUNCHERS:BOOL=1 \
-DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON \
-D Trilinos_SKIP_FORTRANCINTERFACE_VERIFY_TEST:BOOL=TRUE \
-D Trilinos_WARNINGS_AS_ERRORS_FLAGS:STRING="" \
-DTrilinos_ENABLE_TESTS:BOOL=ON \ -DTrilinos_ALLOW_NO_PACKAGES:BOOL=ON \
-DTrilinos_DISABLE_ENABLED_FORWARD_DEP_PACKAGES=ON \
-DTrilinos_DEPS_XML_OUTPUT_FILE:FILEPATH="" \
-DTrilinos_ENABLE_SECONDARY_TESTED_CODE:BOOL=OFF \
-DTrilinos_IGNORE_MISSING_EXTRA_REPOSITORIES:BOOL=ON \
-D Trilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON \
-DCMAKE_BUILD_TYPE:STRING=RELEASE \
-DCOVERAGE_COMMAND:FILEPATH=gcov \
-DMEMORYCHECK_COMMAND:FILEPATH=valgrind \
-DTrilinos_ENABLE_COMPLEX:BOOL=ON \
-DTrilinos_ENABLE_TriKota:BOOL=OFF \
-DCMAKE_VERBOSE_MAKEFILE:BOOL=TRUE \
-DTPL_ENABLE_Matio=OFF \
-DTPL_ENABLE_GLM=OFF \
-DTPL_ENABLE_X11=OFF \
-DIntrepid2_ENABLE_DEBUG_INF_CHECK:BOOL=OFF \
-DPythonInterp_FIND_VERSION=2.7 \
-DTPL_ENABLE_MPI:BOOL=ON \
-DTrilinos_ENABLE_COMPLEX=OFF \
-DTrilinos_TEST_CATEGORIES=BASIC \
-DTrilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON \
-DTrilinos_ENABLE_DEBUG:BOOL=ON \
-DTrilinos_ENABLE_DEBUG_SYMBOLS=OFF \
-DBUILD_SHARED_LIBS:BOOL=ON \
-DTPL_ENABLE_Pthread:BOOL=ON \
-DTPL_ENABLE_Boost:BOOL=ON \
-DTPL_ENABLE_BoostLib:BOOL=ON \
-DTPL_ENABLE_ParMETIS:BOOL=ON \
-DTPL_ENABLE_Zlib:BOOL=ON \
-DTPL_ENABLE_HDF5:BOOL=ON \
-DTPL_ENABLE_Netcdf:BOOL=ON \
-DTPL_BoostLib_INCLUDE_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/boost/1.63.0/gcc/6.1.0/base/include \
-DTPL_BoostLib_LIBRARY_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/boost/1.63.0/gcc/6.1.0/base/lib \
-DTPL_ENABLE_SuperLU:BOOL=ON \
-DAmesos2_ENABLE_KLU2=ON \
-DTeuchos_ENABLE_DEFAULT_STACKTRACE=OFF \
-DTrilinos_TRACE_ADD_TEST=ON \
-DTrilinos_ENABLE_MueLu:BOOL=ON \
-DTrilinos_ENABLE_Stokhos:BOOL=ON \
-DParMETIS_LIBRARY_DIRS=${PARMETIS_DIR}/lib \
-DParMETIS_INCLUDE_DIRS=${PARMETIS_DIR}/include \
-DNetcdf_LIBRARY_DIRS=${NET_CDF_DIR}/lib \
-DNetcdf_INCLUDE_DIRS=${NET_CDF_DIR}/include \
-DSuperLU_LIBRARY_DIRS=${SUPERLU_DIR}/lib \
-DSuperLU_INCLUDE_DIRS=${SUPERLU_DIR}/include \
-DCMAKE_INSTALL_PREFIX:PATH="/scratch/nvrober/install/release" \
$EXTRA_ARGS \
../..
```

**Comment by jjellio** [comment]
@mhoemmen I can test this against the performance build I have on Cori.

**Comment by mhoemmen** [comment]
@CamelliaDPG You've set `Trilinos_ENABLE_DEBUG:BOOL=ON`; this will do a ton of extra debug checks that will definitely slow down Tpetra.  For example, see lines 1732-1470 of Tpetra_MultiVector_def.hpp:
```
#ifdef HAVE_TPETRA_DEBUG
    {
      const bool compat = this->getMap ()->isCompatible (* (A.getMap ()));
      TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(
        ! compat, std::invalid_argument, "Tpetra::MultiVector::dot: *this is "
        "not compatible with the input MultiVector A.  We only test for this "
        "in a debug build.");
    }
#endif //  HAVE_TPETRA_DEBUG
```
Please test a release build with `Trilinos_ENABLE_DEBUG:BOOL=OFF`.  I also recommend testing with Intel 17, since we know that this does a better job of inlining through Kokkos::View::operator().

**Comment by jjellio** [comment]
Cori has been down off and on for a few days (dedicated runs, upgrades, etc..)

My build is KNL, with the default execution space set to OpenMP. Also, I linked against MKL v17 which is also threaded. The compiler here is Intel 17, used under Cray's environment (static linkage, etc...)

My Trilinos source was pulled around June 9th I believe, but that could be wrong.

Because of the above, I configure the environment as:

```
export OMP_NUM_THREADS=1
export OMP_PLACES=cores
export OMP_PROC_BIND=spread
CMD=./epetra-test.exe


srun --nodes=1 --ntasks-per-node=1 --cpu_bind=cores ${CMD} 1000000 50 |& tee -a epetra.log
srun --nodes=1 --ntasks-per-node=1 --cpu_bind=cores ${CMD} 100000 500 |& tee -a epetra.log
srun --nodes=1 --ntasks-per-node=1 --cpu_bind=cores ${CMD} 10000 5000 |& tee -a epetra.log
srun --nodes=1 --ntasks-per-node=1 --cpu_bind=cores ${CMD} 1000 50000 |& tee -a epetra.log
srun --nodes=1 --ntasks-per-node=1 --cpu_bind=cores ${CMD} 100 500000 |& tee -a epetra.log
srun --nodes=1 --ntasks-per-node=1 --cpu_bind=cores ${CMD} 10 5000000 |& tee -a epetra.log
srun --nodes=1 --ntasks-per-node=1 --cpu_bind=cores ${CMD} 1 50000000 |& tee -a epetra.log
```

Results 

  numDots |  globalRowCount |  Epetra |  Tpetra |  Ratio
 ------------- | --- | --- | --- | ---
  1000000 |  50 |  1.482 |  57.53 |  38.82
  100000 |  500 |  0.1678 |  5.761 |  34.33
  10000 |  5000 |  0.05415 |  0.6191 |  11.43
  1000 |  50000 |  0.0954 |  0.1702 |  1.78
  100 |  500000 |  0.09273 |  0.08578 |  0.93  
  10 |  5000000 |  0.09347 |  0.07492 |  0.80  
  1 |  50000000 |  0.09332 |  0.07352 |  0.79 

Obviously, something isn't optimal at small vector lengths. I believe dot() is a Kokkos implementation of ddot() for Tpetra. Does Epetra call Teuchos::BLAS? If this is the case, I find it interesting that Tpetra's Kokkos loop is outperforming MKL at large sizes.

It also would be worth redoing the above with Tpetra's node type set to Serial.

**Comment by CamelliaDPG** [comment]
@mhoemmen Ah, sorry about that!  I must have changed to debug a while back—since I have this in a build directory called "release", it didn't occur to me that I might be running a debug build.  I am now rebuilding (with gcc-6 to start) and will redo the tests after that.  (FWIW, I started down this road due to some profiling results which used a non-debug build of Trilinos—on those runs, the vectors were of length O(1000), so I'm glad to see that that @jjellio's testing confirms that there is some sub-optimality  at around that length.  Also glad to see that it does not appear to affect longer vectors in the same way—but I will confirm this for my setup as well.)

**Comment by CamelliaDPG** [comment]
@mhoemmen @jjellio Thanks for looking at this!

**Comment by jhux2** [comment]
> Ah, sorry about that! I must have changed to debug a while back—since I have this in a build directory called "release", it didn't occur to me that I might be running a debug build.

@CamelliaDPG Fwiw, this has bitten me more times than I care to admit.  I finally put a check in my base configure script that all my other scripts call to help catch such cases.

```
if [ -z ${BUILD_MODE+x} ]; then
  echo "$0: BUILD_MODE has not been specified in the cmake configure options."
  echo "Did you remember to \"export\" the variable?"
  exit
fi

if [ -e CMakeCache.txt ]; then
  # Error if the current Cmake build type differs from what was done previously
  lastBuildType=`egrep "CMAKE_BUILD_TYPE:" CMakeCache.txt`
  matchMode=`echo ${lastBuildType} | grep ${BUILD_MODE}`
  if [ -z $matchMode ]; then
    echo "Requested build type \"${BUILD_MODE}\" doesn't match previous configuration (${lastBuildType})."
    echo "If this is what you intended, delete the CMakeCache.txt file and reconfigure."
    exit
  fi

  lastCompiler=`egrep "CMAKE_CXX_COMPILER:" CMakeCache.txt`
  thisCompiler=`which mpicxx`
  matchMode=`echo ${lastCompiler} | grep ${thisCompiler}`
  if [ -z $matchMode ]; then
    lastVer=`echo $lastCompiler | cut -d= -f 2`
    echo "Currently loaded compiler version \"${thisCompiler}\" doesn't match previous
configuration"
    echo "(${lastVer})."
    echo "If this is what you intended, delete the CMakeCache.txt file and reconfigure."
    exit
  fi

  lastSource=`egrep "Trilinos_SOURCE_DIR:STATIC=" CMakeCache.txt`
  matchMode=`echo ${lastSource} | grep ${TRILINOS_HOME}`
  if [ -z $matchMode ]; then
    lastSource=`echo $lastSource | cut -d= -f 2`
    echo "Current Trilinos source \"${TRILINOS_HOME}\" doesn't match previous source"
    echo "(${lastSource})"
    echo "If this is what you intended, delete the CMakeCache.txt file and reconfigure."
    exit
  fi


fi
```

**Comment by mhoemmen** [comment]
Thanks @jjellio ! :-D

@jjellio wrote:
> I believe dot() is a Kokkos implementation of ddot() for Tpetra. 

Yes; Tpetra uses `KokkosBlas::dot`.

> Obviously, something isn't optimal at small vector lengths. 

This is likely OpenMP kernel launch overhead.  We could optimize by dispatching to the `Kokkos::Serial` specialization if the vectors are too short.

> Does Epetra call `Teuchos::BLAS`? 

Epetra calls the BLAS for dot products, through the `Epetra_BLAS` interface.  `Teuchos::BLAS` was modeled after `Epetra_BLAS`.

If you look at the code in `Epetra_MultiVector`, there is an OpenMP option with a hand-rolled kernel that doesn't call the BLAS.  However, that option is disabled by default.  Epetra's default use of OpenMP only does computations that will not cause floating-point differences due to parallelism.  Dot products and norms are excluded by default.

**Comment by mhoemmen** [comment]
@jjellio wrote:
> It also would be worth redoing the above with Tpetra's node type set to Serial.

This would be a good way to exclude OpenMP kernel launch overhead.

**Comment by jjellio** [comment]
@mhoemmen Best thing to do is to change this to Serial, then run it under a profiler (maybe I can do that when I get out of report writing hell). The overhead of checking Kokkos view stuff could be relatively large at such small vector lengths.

**Comment by mhoemmen** [comment]
@jjellio Maybe we need to get the Kokkos folks involved.  There could be a short loop count optimization that we could pass along to them.

**Comment by nmhamster** [comment]
@jjellio @mhoemmen  - have you thought about just calling a direct OpenMP implementation (like we do in SpMV) over a Kokkos implementation? Kokkos is optimized around larger parallel calls, the OpenMP can usually handle these better. Plus, I suspect this code is not vectorizing where direct OpenMP may well.

**Comment by mhoemmen** [comment]
@nmhamster This question may sound antagonistic: Are we supposed to assume that compilers can't vectorize through Kokkos::parallel_for? Also, how large is large here?  The crossover point was 50k rows; is that big by OpenMP standards?

**Comment by jjellio** [comment]
@mhoemmen This seems to be an issue with the layers between Tpetra::dot() and the actual dot kernel.  Serial or OpenMP is not important.  Vtune shows a substantial amount of time in the constructor for Tpetra::Vector().

The call tree looks like:
MV::dot() -> Tpetra::Vector (constructor) -> Tpetra::MV() (constructor) -> Tpetra::DistObject.  The bulk of the time is the in DistObject. DistObject then spends its time making a DualView.

The other badness is the destructor for the Vector, which is then spending time in the DistObject destructor.

The short version seems to be that the cost of getting a Tpetra::Vector from a MV is not cheap. Or is the real issue that getVector() should not be triggering this expensive constructor?

**Comment by mhoemmen** [comment]
@jjellio That `Tpetra::Vector` constructor just makes a view of the first column.  It shouldn't be expensive.  If it is, that's bad, because it's just calling `Kokkos::subview` a couple times.

**Comment by jjellio** [comment]
@mhoemmen The problem seems related to Tpetra::Map()'s copy constructor (or lack thereof). I could be missing something, but when getVector (or getVecNonConst) is called, this ends up with DistObject calling map_ (map).  Which I believe should be a map shallow copy, but map does not provide a copy constructor. Something isn't right with the default copy constructor for Map. This manifests as distObject's constructor spending a large amount of time in DualView() construction.  What isn't obvious is what is wrong with the default constructor.  By definition, Kokkos views should be shallow copied.

**Comment by mhoemmen** [comment]
@jjellio wrote:
> DistObject calling map_ (map). Which I believe should be a map shallow copy, but map does not provide a copy constructor.

DistObject's `map_` field is an `RCP<const map_type>`.  The code `map_(map)` should just shallow-copy the RCP.

**Comment by mhoemmen** [comment]
@jjellio `map_ (map)` invokes the RCP copy constructor, not Map's copy constructor.  Map doesn't store a DualView.  If you see three or four DualViews getting created, then it's DistObject's constructor (that takes a Map).  We can fix this by avoiding the getVector() call.  I have a fix in progress (I'll open an issue to describe it).

**Comment by mhoemmen** [comment]
#1560 should help with this, by reducing getVector() overhead.

**Comment by mhoemmen** [comment]
See also #1470.
